accepted/engine-host-functions.md

```diff
@@ -0,0 +1,211 @@
+
+This proposal calls for adding the following methods to `Store`:
+ 
+```rust
+/// Gets a function from the `Engine` associated with this `Store`.
+///
+/// Returns `None` if the given function is not defined.
+pub fn get_engine_function(&self, module: &str, name: &str) -> Option<Func>;
+
+/// Get a reference to a context value of a particular type.
+///
+/// Returns `None` if the context value of the given type does not exist.
+pub fn get_context<T: Any>(&self) -> Option<&T>;
+ 
+/// Sets a context value.
+///
+/// Returns the previous context value of the same type if present.
+pub fn set_context<T: Any>(&mut self, value: T) -> Option<T>;
+ 
+/// Removes a context value.
+///
+/// Returns `None` if a context value of the given type does not exist.
+pub fn remove_context<T: Any>(&mut self) -> Option<T>;
+```
+ 
+`get_engine_function`  will register the function's instance handle with the `Store`, but as a *borrowed* handle which it is not responsible for deallocation. As `Store` keeps a reference on its associated `Engine`, this ownership model should be sufficient to prevent `funcref` values or imports of engine functions from outliving the function instance itself.
+
+For host functions that require contextual data (e.g. WASI), the `set_context` method will be used for associating context with a `Store` and it can later be retrieved via `caller.store().get_context()`.
+ 
+## Changes to `wasmtime::Linker`
+ 
+There will be no changes to the API surface of `Linker` to support the changes proposed in this RFC.
+ 
+However, the `Linker` implementation will be changed such that it will fallback to calling `Store::get_engine_function` when resolving imports for `Linker::instantiate` and `Linker::module`.
+ 
+This should allow for overriding engine function imports in the `Linker` if an import of the same name has already been defined in the `Engine`.
+ 
+## The `wasmtime::IntoEngineFunc` trait
+ 
+Much like `IntoFunc` that is used to easily define functions using `Func::wrap`, an `IntoEngineFunc` trait will be implemented for `Fn([Caller], P1, P2...) -> R` for up to sixteen parameters that implement `WasmTy` and the return type implements `WasmRet`.
+ 
+`IntoEngineFunc` will also be implemented for ``Fn(Caller, &[Val], &mut [Val]) -> Result<Trap> + `static`` so that any number of parameters and return values can be supported for engine functions.
+ 
+This trait, like `IntoFunc`, is intended as an internal implementation detail and will not be documented.
+ 
+## Changes to `wasmtime_wasi::Wasi`
+ 
+`Wasi` is a type generated from `wasmtime-wiggle`.
+ 
+This proposal adds the following methods to the generated `Wasi` type:
+ 
+```rust
+/// Adds WASI functions to the given `Engine`.
+pub fn add_to_engine<T>(engine: &Engine) -> Result<()>;
+
+/// Sets the given `WasiCxt` for the `Store`.
+///
+/// This will replace any existing `WasiCtx` associated with the `Store`.
+pub fn set_context(store: &Store, context: WasiCtx);
```

Comment:
```suggestion
/// This will replace any existing `WasiCtx` associated with the `Store`.
///
/// The old `WasiCtx`, if any, is returned.
pub fn set_context(store: &Store, context: WasiCtx) -> Option<WasiCtx>;
```

Although, will this overwrite *any* context that already exists on the store? Not just a `WasiCtx`?

This is why it would be nice if we used [ephemerons](https://en.wikipedia.org/wiki/Ephemeron) for associating context instead. (Something we can always add later, if it makes sense; nothing to block this stuff on I don't think.)

Reply:
It would just replace the `WasiCtx` as a map of typeid -> context value will be used ala Lucet's `set_embed_ctx`.